### ISSUE TITLE
Request headers are overriden when using `LDflex-Comunica`

### ISSUE THREAD

**Issue by NSeydoux** [inrupt/solid-auth-fetcher#88]
# Description of the bug
 
When called with the following values:
```
"init": {
  "headers": {
    "_headers": {
      "accept": [
        "application/n-quads,application/trig;q=0.95,application/ld+json;q=0.9,application/n-triples;q=0.8,text/turtle;q=0.5,application/rdf+xml;q=0.5,application/json;q=0.45,text/n3;q=0.35,application/xml;q=0.3,text/xml;q=0.3,image/svg+xml;q=0.3,text/html;q=0.2,application/xhtml+xml;q=0.18"
      ]
    }
  }
},
"input": "https://www.w3.org/People/Berners-Lee/card"
```
in a `NodeJS` environment (using `cross-fetch)`, `solid-auth-fetcher` performs a query with default `Accept` headers, instead of the ones specified in the `init` data structure. 

This bug was observed while upgrading [LDflex-Comunica](https://github.com/NSeydoux/LDflex-Comunica). I'm not entirely sure the `init` structure above is compatible with the expected `RequestInit` [type](https://microsoft.github.io/PowerBI-JavaScript/interfaces/_node_modules_typedoc_node_modules_typescript_lib_lib_dom_d_.requestinit.html), namely the additional wrapping `_headers` seems extraneous. However, this exact same structure was processed successfully by `solid-auth-client`. A special case might be added to patch the issue on a short term, but if the problem comes from the data structure passed by `LDflex-Comunica`, then actual patch should be implemented there.

# Steps to reproduce

Clone `https://github.com/NSeydoux/LDflex-Comunica`, run `npm ci` and `npm run test`. The tests timeout for a different reason, but the core issue is that the request isn't stubbed by PollyJS because it doesn't match the expected signature because the headers have been mangled.

**Comment by RubenVerborgh** [comment]
Well the `_headers` structure definitely isn't standard, but I presume it is a flattened version of an object conforming to the `Headers` interface (https://developer.mozilla.org/en-US/docs/Web/API/Headers).

**Comment by jaxoncreed** [comment]
Ah I think this is an auth fetcher problem. Specifically looking at this line https://github.com/inrupt/solid-auth-fetcher/blob/master/src/authenticatedFetch/dpop/DpopAuthenticatedFetcher.ts#L70. The init headers should be transferred here.

**Comment by RubenVerborgh** [comment]
The problem is that auth fetcher assumes in various places that `headers` is an object literal, whereas it can actually be one of three:

> headers: Any headers you want to add to your request, contained within a Headers object or an object literal with ByteString values.

—https://developer.mozilla.org/en-US/docs/Web/API/Request/Request

So we need a method `normalizeHeaders` or so, that converts them into a single format, for easier dealing below. I suggest that format to be `Headers`, because it deals with casing differences etc.

The first bug is at

https://github.com/inrupt/solid-auth-fetcher/blob/4fac0701e34f28ae4b3809bfcf8f1aa0172002cd/src/authenticatedFetch/dpop/DpopAuthenticatedFetcher.ts#L78-L80

where the object literal assumption is made, and then the second one at

https://github.com/inrupt/solid-auth-fetcher/blob/4fac0701e34f28ae4b3809bfcf8f1aa0172002cd/src/authenticatedFetch/dpop/DpopAuthenticatedFetcher.ts#L78-L80

where that same assumption is made, in addition to the assumption that headers will always be spelled lowercase (which they are not).

TL;DR: convert to `Headers` before manipulating them.

**Comment by RubenVerborgh** [comment]
Might even be sufficient to just call the `Headers` constructor (https://developer.mozilla.org/en-US/docs/Web/API/Headers/Headers), although it doesn't mention compatibility with object literals (but the examples do).

**Comment by NSeydoux** [comment]
There is something very weird happening: 
```
"headers": {
    "_headers": {
      "accept": [
        "application/n-quads,application/trig;q=0.95,application/ld+json;q=0.9,application/n-triples;q=0.8,text/turtle;q=0.5,application/rdf+xml;q=0.5,application/json;q=0.45,text/n3;q=0.35,application/xml;q=0.3,text/xml;q=0.3,image/svg+xml;q=0.3,text/html;q=0.2,application/xhtml+xml;q=0.18"
      ]
    }
  }
```
is actually a valid result for `JSON.stringify(someHeader, null, "  ")`. In the case of `LDflex-Comunica`, the tests run unauthenticated fetches, so the faulty call in `solid-auth-fetcher` is indeed https://github.com/inrupt/solid-auth-fetcher/blob/4fac0701e34f28ae4b3809bfcf8f1aa0172002cd/src/authenticatedFetch/dpop/DpopAuthenticatedFetcher.ts#L78-L80 .

But replacing `headers: { ...requestInitiWithDefaults.headers }` with `headers: { new Headers(requestInitiWithDefaults.headers) }` does not resolved the issue, since the resulting object will still be eventually stringified in the same way by `node-fetch`, which seems to assume there is not intermediary `_headers` object between the `headers` object and the actual values. Therefore, a (non-type safe, local test-only) fix is to change `headers: { ...requestInitiWithDefaults.headers }` to `headers: requestInitiWithDefaults.headers._headers`. 

For the time being, a fix that works is to flatten the object into a `Record<string, string>`: 
```
if (!authToken) {
      const flatHeaders: Record<string, string> = {};
      const headersToFlatten = new Headers(requestInitiWithDefaults.headers);
      headersToFlatten.forEach((value, key) => {
        flatHeaders[key] = value;
      });
      // perform unauthenticated fetch
      return this.fetcher.fetch(url, {
        ...requestInit,
        headers: flatHeaders
      });
    }
``` 

i'll submit the associated PR.

**Comment by RubenVerborgh** [comment]
> is actually a valid result for `JSON.stringify(someHeader, null, " ")`.

Yes, that's what happens when you _flatten_ one particular implementation of `Headers`. However, when flattening, you lose its methods etc., which expose the _internal_ `_headers` structure.

> since the resulting object will still be eventually stringified in the same way by node-fetch,

So the question is whether node-fetch a) is correctly passed a `Headers` object b) works with `Headers` objects.

But don't let the `_headers` distract us; it's an internal field from some `Headers` implementation that we should not worry about or access in any way.

**Comment by RubenVerborgh** [comment]
And if node-fetch does not support it, that's a bug we should report over there, but we can simply flatten them by `headers: Object.fromEntries(headers.entries())`
